src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
+            comp.VerifyDiagnostics(
+                // (3,19): error CS8936: Feature 'ref fields' is not available in C# 10.0. Please use language version 11.0 or greater.
+                // var r = new R() { field = ref x }; // 1
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion10, "field").WithArguments("ref fields", "11.0").WithLocation(3, 19),
+                // (6,15): error CS8936: Feature 'ref fields' is not available in C# 10.0. Please use language version 11.0 or greater.
+                // _ = r2 with { field = ref x }; // 2
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion10, "field").WithArguments("ref fields", "11.0").WithLocation(6, 15)
+                );
+        }
+
+        [Fact]
+        public void RefInitializer()
+        {
+            var source = @"
+public class C
+{
+    public static void Main()
+    {
+        int x = 42;
+        var r = new R() { field = ref x };
+        System.Console.Write(r.ToString());
+    }
+}
+
+ref struct R
+{
+    public ref int field;
+    public override string ToString()
+    {
+        return field.ToString();
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(source, verify: Verification.Skipped, expectedOutput: IncludeExpectedOutput("42"));
+            verifier.VerifyIL("C.Main",
+"""
+{
+  // Code size       40 (0x28)
+  .maxstack  2
+  .locals init (int V_0, //x
+                R V_1, //r
+                R V_2)
+  IL_0000:  ldc.i4.s   42
+  IL_0002:  stloc.0
+  IL_0003:  ldloca.s   V_2
+  IL_0005:  initobj    "R"
+  IL_000b:  ldloc.2
+  IL_000c:  ldfld      "ref int R.field"
+  IL_0011:  ldloc.0
+  IL_0012:  stind.i4
+  IL_0013:  ldloc.2
+  IL_0014:  stloc.1
+  IL_0015:  ldloca.s   V_1
+  IL_0017:  constrained. "R"
+  IL_001d:  callvirt   "string object.ToString()"
+  IL_0022:  call       "void System.Console.Write(string)"
+  IL_0027:  ret
+}
+""");
+        }
+
+        [Fact]
+        public void RefInitializer_RHSMustBeDefinitelyAssigned()
+        {
+            // The right operand must be definitely assigned at the point of the ref assignment.
+            var source = @"
+int x;
+var r = new R() { field = ref x };
+
+ref struct R
+{
+    public ref int field;
+    public override string ToString()
+    {
+        return field.ToString();
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (3,31): error CS0165: Use of unassigned local variable 'x'
+                // var r = new R() { field = ref x };
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(3, 31)
+                );
+        }
+
+        [Fact]
+        public void RefInitializer_RHSTypeMustMatch()
+        {
+            // The right operand must be an expression that yields an lvalue designating a value of the same type as the left operand.
```

Comment:
consider testing a case where there is an implicit conversion from the RHS type to the LHS type.